### PR TITLE
8281096: Flags introduced by configure script are not passed to ADLC build

### PR DESCRIPTION
**Pull request by gnu-andrew**
The 8u configure script defines compiler flags in `EXTRA_CFLAGS`, `EXTRA_LDFLAGS` and `EXTRA_ASFLAGS`. Some are added by configure tests, while others are taken directly from corresponding options passed by the user.

8u still use the legacy HotSpot build system which is not fully integrated with the autoconf system. Variables defined by configure thus have to be explicitly passed down to the separate HotSpot build.

ADLC is a tool used at build-time and so the flags it uses don't impact on the end product. So, for a long time, it has been ignoring these flags defined by configure and using just its own minimal set.

However, with newer compilers, this means that the code is compiled to a newer version of the C++ standard, as the default has changed in GCC 6 and later (see [JDK-8151841](https://bugs.openjdk.org/browse/JDK-8151841)). With the latest versions of GCC (11 and 12), this actually leads to build failures due to the use of 'register' (GCC 11) and the way comments are used (GCC 12) in the code.

We should fix the ADLC build to use the same flags as the rest of the build. The impact should be negligible, given the same flags are already used in the code that is actually shipped.

This does not affect 9+ where HotSpot's build system has been replaced with full integration in the autoconf system.

With this change, 8u can be built with GCC 11 on GNU/Linux. I'd appreciate testing on other platforms, particularly those not covered by GHA (Solaris, AIX - @adamfarley, @sxa & @deepa181 who have provided previous fixes for these platforms)

During testing, we found we could not use the `EXTRA_` prefixed variables as they include flags for the target compiler on cross-compilation builds. To fix this, we split out new `HOST_` prefixed variables which include all the flags of the `EXTRA_` prefixed variables, bar the `--sysroot` options.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] [JDK-8281096](https://bugs.openjdk.org/browse/JDK-8281096) needs maintainer approval
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8281096](https://bugs.openjdk.org/browse/JDK-8281096): Flags introduced by configure script are not passed to ADLC build (**Bug** - P4 - Approved)

### Reviewers
 * [Thomas Stuefe](https://openjdk.org/census#stuefe) (@tstuefe - **Reviewer**)
 * [Zdenek Zambersky](https://openjdk.org/census#zzambers) (@zzambers - Committer)
 * [Severin Gehwolf](https://openjdk.org/census#sgehwolf) (@jerboaa - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk8u-dev.git pull/357/head:pull/357` \
`$ git checkout pull/357`

Update a local copy of the PR: \
`$ git checkout pull/357` \
`$ git pull https://git.openjdk.org/jdk8u-dev.git pull/357/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 357`

View PR using the GUI difftool: \
`$ git pr show -t 357`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk8u-dev/pull/357.diff">https://git.openjdk.org/jdk8u-dev/pull/357.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk8u-dev/pull/357#issuecomment-1689601278)

### CODE REVIEW

**Review by tstuefe — COMMENTED** [review]
Looks good and makes sense. Why not passing LFLAGS to solaris?

**Comment by gnu-andrew** [comment]
/open

@deepa181 thanks for the feedback. It's good to know it doesn't break AIX.
@tstuefe good catch. I need to find out why `LFLAGS` is missing on Solaris (assuming at the moment it's not used generally) and also check why it seemed to be failing on cross-compiles (we don't use them ourselves so that is untested prior to this)

**Comment by gnu-andrew** [comment]
> Looks good and makes sense. Why not passing LFLAGS to solaris?

It seems this matches what Solaris does in `vm.make`. These new lines are effectively just copied over from `vm.make` which handles the rest of the build, and for some reason, Solaris does not set `LFLAGS` for the VM build either.

**Comment by gnu-andrew** [comment]
I think I know what's wrong here now with the cross-compilation builds. The `EXTRA_CFLAGS` that HotSpot gets end up including the `--sysroot` options on cross-compile builds, which are really destined for the target compiler, not a build compiler which is compiling a build tool that is going to be run during the build.

I can fix this, but it's going to make this fix a bit more involved than I initially thought.

**Comment by sxa** [comment]
@gnu-andrew Finally got around to looking at this (and this comment will stop the bot trying to close it!) and performed an initial test on Linux/aarch64 before trying elsewhere. I can build ok with GCC 10.3 but not 11.2 due to this error:

`/home/sxa/temurin-build/build-farm/workspace/build/src/hotspot/src/share/vm/opto/type.cpp:2556:71: error: 'this' pointer is null [-Werror=nonnull]`

**Comment by sxa** [comment]
Solaris build of this branch using our default Sun Studio compiler seems to build ok (x64 and SPARC) 👍🏻

**Comment by gnu-andrew** [comment]
> @gnu-andrew Finally got around to looking at this (and this comment will stop the bot trying to close it!) and performed an initial test on Linux/aarch64 before trying elsewhere. I can build ok with GCC 10.3 but not 11.2 due to this error:
> 
> `/home/sxa/temurin-build/build-farm/workspace/build/src/hotspot/src/share/vm/opto/type.cpp:2556:71: error: 'this' pointer is null [-Werror=nonnull]`

Thanks for verifying this. Once we can get this change in, I'll look and replicating other failures and resolving those. I think the `type.cpp` one has been raised in a few other places, but I've not seen it myself (maybe it requires warnings as errors)

**Comment by gnu-andrew** [comment]
> _Mailing list message from [Thorsten Glaser](mailto:t.glaser@qvest-digital.com) on [jdk8u-dev](mailto:jdk8u-dev@mail.openjdk.org):_
> 
> On Mon, 19 Feb 2024, Andrew John Hughes wrote:
> 
> > I think I know what's wrong here now with the cross-compilation builds.
> > The `EXTRA_CFLAGS` that HotSpot gets end up including the `--sysroot`
> > options on cross-compile builds, which are really destined for the
> > target compiler, not a build compiler which is compiling a build tool
> > that is going to be run during the build.
> 
> Ouch. So no clean distinction between CFLAGS and HOSTCFLAGS?

Yes, this is the conclusion I've come to as well.

> 
> > I can fix this, but it's going to make this fix a bit more involved
> > than I initially thought.
> 
> Yocto has an ugly workaround: https://github.com/ostroproject/ostro-os/blob/master/meta-java/recipes-core/openjdk/patches-openjdk-8/openjdk8-fix-adlc-flags.patch

Ah good to see someone else has encountered this. I think we can resolve it in a similar way, but by defining a new flag in configure rather than trying to filter things apart later.

> 
> bye, //mirabilos -- Infrastrukturexperte ? Qvest Digital AG Am Dickobskreuz 10, D-53121 Bonn ? https://www.qvest-digital.com/ Telephon +49 228 54881-393 ? Fax: +49 228 54881-235 HRB AG Bonn 18196 ? USt-ID (VAT): DE274355441 Vorstand: Dr. Stefan Barth, Kai Ebenrett, Boris Esser, Alexander Steeg Vorsitzender Aufsichtsrat: Peter N?then

**Comment by gnu-andrew** [comment]
I've rewritten the patch so it now uses `HOST_CFLAGS` and `HOST_LDFLAGS` in the ADLC build. These are created in `configure` by not including the `sysroot` arguments.

The below shows the old and new flags in the aarch64 cross-compile, which now passes:

~~~
# Retain EXTRA_{CFLAGS,CXXFLAGS,LDFLAGS,ASFLAGS} for the target flags to
# maintain compatibility with the existing Makefiles
EXTRA_CFLAGS= --sysroot="/home/runner/sysroot-arm64/" -fstack-protector  $(NO_DELETE_NULL_POINTER_CHECKS_CFLAG) \
				    $(NO_LIFETIME_DSE_CFLAG) $(CXXSTD_CXXFLAG)
EXTRA_CXXFLAGS= --sysroot="/home/runner/sysroot-arm64/" -fstack-protector 
EXTRA_LDFLAGS= --sysroot="/home/runner/sysroot-arm64/" -Wl,-z,relro 
EXTRA_ASFLAGS=
# Define an equivalent set for the host flags (i.e. without sysroot options)
HOST_CFLAGS= -fstack-protector  $(NO_DELETE_NULL_POINTER_CHECKS_CFLAG) \
				 $(NO_LIFETIME_DSE_CFLAG) $(CXXSTD_CXXFLAG)
HOST_CXXFLAGS= -fstack-protector 
HOST_LDFLAGS= -Wl,-z,relro
HOST_ASFLAGS=
~~~

I'll remove the verbose build log now, but leave the printing of the spec files as I think it does no harm and may be useful for future debugging.

The remaining failures are due to download issues (Debian Buster release on some architectures, Windows VS2010) and the cacerts tests, both of which are not caused by this patch. It may be worth trying to bump the cross-compile to a later Debian release in another PR, as Buster is about to go EOL.

**Comment by gnu-andrew** [comment]
> I'll remove the verbose build log now, but leave the printing of the spec files as I think it does no harm and may be useful for future debugging.

Actually looks like the main build has had `LOG_LEVEL=debug` since inception so I'll leave this in for cross-compiles. It makes things easier to debug.

**Review by tstuefe — COMMENTED** [review]
Mostly good. Weird test errors in GHA, though.

File: .github/workflows/submit.yml [hunk 1]
```diff
@@ -452,7 +452,10 @@ jobs:
         working-directory: jdk
 
       - name: Build
-        run: make CONF_NAME=linux-${{ matrix.gnu-arch }}-hotspot hotspot
+        run: |
+          cat build/linux-${{ matrix.gnu-arch }}-hotspot/spec.gmk
+          cat build/linux-${{ matrix.gnu-arch }}-hotspot/hotspot-spec.gmk
+          make CONF_NAME=linux-${{ matrix.gnu-arch }}-hotspot LOG_LEVEL=debug hotspot
```
[the remark this answers is not in the document]
Comment (gnu-andrew) [reply .github/workflows/submit.yml:458]: https://github.com/openjdk/jdk8u-dev/pull/357#issuecomment-2125992116

Yeah. They are not needed, but I see no harm in leaving them in. The output is small in comparison to the whole log and it's useful to have if the build does fail. We've had similar in the RPM spec files for a long time.

**Comment by gnu-andrew** [comment]
> Mostly good. Weird test errors in GHA, though.

Thanks for looking over it again. Which ones do you mean? The download issues and cacert failures aren't related to this change. I seem them on other PRs.

**Review by tstuefe — APPROVED** [review]
@gnu-andrew Okay, if you say the GHA failures are unrelated to this patch, this is okay for me.

**Comment by gnu-andrew** [comment]
/approval request Fixes ADLC build with GCC 11+ by passing through compiler flags used for the rest of the HotSpot build. Later versions of OpenJDK are not affected, due to the HotSpot build being integrated into the new build system there.

**Comment by gnu-andrew** [comment]
Merged with latest jdk8u-dev to bring in @zzambers ' changes to the Windows & alt-Linux builds. All builds now pass and the only test failure seems unrelated (`com/sun/jdi/JdbExprTest.sh`)

**Comment by zzambers** [comment]
@gnu-andrew I think remaining failure should get fixed by: https://github.com/openjdk/jdk8u-dev/pull/497

**Comment by gnu-andrew** [comment]
> @gnu-andrew I think remaining failure should get fixed by: #497

Ah, thanks for spotting that one and getting it sponsored.

**Comment by zzambers** [comment]
I have done some additional testing:
- I  [tested](https://github.com/zzambers/jdk8u-dev/actions/runs/9664993178)  this, with GHA workflows updated to ubuntu-22.04 on top (+ [fix for jdi test](https://github.com/openjdk/jdk8u-dev/pull/497)). This is because, I have recently seen build failures on some archs, when I [tried to update GHA testing to ubuntu-22.04](https://github.com/openjdk/jdk8u-dev/pull/491#issue-2281128235), while looking at another issue. There are no build failures with this changeset (1 test failure/error on linux x86 seems unrelated). 

- I have also tried to do build in rhel-9 VM (gcc-11.4.1). It passed ADLC build (unlike unpatched version). Build later failed on another (separate) error in `hotspot/src/share/vm/opto/type.cpp` (see https://github.com/openjdk/jdk8u-dev/pull/479#issuecomment-2058078166). If I also modified `type.cpp` (based on that PR), build passed. 

Change looks good. (works for me)

**Review by zzambers — APPROVED** [review]

**Review by jerboaa — APPROVED** [review]

**Comment by gnu-andrew** [comment]
> I have done some additional testing:
> 
>     * I  [tested](https://github.com/zzambers/jdk8u-dev/actions/runs/9664993178)  this, with GHA workflows updated to ubuntu-22.04 on top (+ [fix for jdi test](https://github.com/openjdk/jdk8u-dev/pull/497)). This is because, I have recently seen build failures on some archs, when I [tried to update GHA testing to ubuntu-22.04](https://github.com/openjdk/jdk8u-dev/pull/491#issue-2281128235), while looking at another issue. There are no build failures with this changeset (1 test failure/error on linux x86 seems unrelated).
> 
>     * I have also tried to do build in rhel-9 VM (gcc-11.4.1). It passed ADLC build (unlike unpatched version). Build later failed on another (separate) error in `hotspot/src/share/vm/opto/type.cpp` (see [8329826: GCC 12 reports some compiler error when building jdk8 #479 (comment)](https://github.com/openjdk/jdk8u-dev/pull/479#issuecomment-2058078166)). If I also modified `type.cpp` (based on that PR), build passed.
> 
> 
> Change looks good. (works for me)

Thanks for testing. Can you give me more details on what kind of build failed in `type.cpp` (e.g. release/slowdebug/fastdebug)?  With just this patch, it builds for me on GCC 13.

**Comment by gnu-andrew** [comment]
Thanks for the approval, @jerboaa 
/integrate

**Comment by zzambers** [comment]
> > I have done some additional testing:
> > ```
> > * I  [tested](https://github.com/zzambers/jdk8u-dev/actions/runs/9664993178)  this, with GHA workflows updated to ubuntu-22.04 on top (+ [fix for jdi test](https://github.com/openjdk/jdk8u-dev/pull/497)). This is because, I have recently seen build failures on some archs, when I [tried to update GHA testing to ubuntu-22.04](https://github.com/openjdk/jdk8u-dev/pull/491#issue-2281128235), while looking at another issue. There are no build failures with this changeset (1 test failure/error on linux x86 seems unrelated).
> > 
> > * I have also tried to do build in rhel-9 VM (gcc-11.4.1). It passed ADLC build (unlike unpatched version). Build later failed on another (separate) error in `hotspot/src/share/vm/opto/type.cpp` (see [8329826: GCC 12 reports some compiler error when building jdk8 #479 (comment)](https://github.com/openjdk/jdk8u-dev/pull/479#issuecomment-2058078166)). If I also modified `type.cpp` (based on that PR), build passed.
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Change looks good. (works for me)
> 
> Thanks for testing. Can you give me more details on what kind of build failed in `type.cpp` (e.g. release/slowdebug/fastdebug)? With just this patch, it builds for me on GCC 13.

I can still reproduce issue with `type.cpp` on rhel-9, with just most default build (just boot jdk set). Gcc version is 11.4.1 (system one).

Errors look following way (see [jdk8-type.cpp-error.log](https://github.com/user-attachments/files/16615930/jdk8-type.cpp-error.log) for full configure/make log):
```
In constructor ‘TypeOopPtr::TypeOopPtr(Type::TYPES, TypePtr::PTR, ciKlass*, bool, ciObject*, int, int, const TypeOopPtr*, int)’,
    inlined from ‘TypeAryPtr::TypeAryPtr(TypePtr::PTR, ciObject*, const TypeAry*, ciKlass*, bool, int, int, bool, const TypeOopPtr*, int)’ at /home/tester/jdk8u-dev/hotspot/src/share/vm/opto/type.hpp:1094:39,
    inlined from ‘static const TypeAryPtr* TypeAryPtr::make(TypePtr::PTR, const TypeAry*, ciKlass*, bool, int, int, const TypeOopPtr*, int)’ at /home/tester/jdk8u-dev/hotspot/src/share/vm/opto/type.cpp:3745:115:
/home/tester/jdk8u-dev/hotspot/src/share/vm/opto/type.cpp:2556:71: error: ‘this’ pointer is null [-Werror=nonnull]
 2556 |           ciInstanceKlass* k = o->as_instance()->java_lang_Class_klass()->as_instance_klass();
      |                                ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~
In file included from /home/tester/jdk8u-dev/hotspot/src/share/vm/ci/ciCallSite.hpp:28,
                 from /home/tester/jdk8u-dev/hotspot/src/share/vm/code/dependencies.hpp:28,
                 from /home/tester/jdk8u-dev/hotspot/src/share/vm/ci/ciEnv.hpp:32,
                 from /home/tester/jdk8u-dev/hotspot/src/share/vm/ci/ciUtilities.hpp:28,
                 from /home/tester/jdk8u-dev/hotspot/src/share/vm/ci/ciNullObject.hpp:30,
                 from /home/tester/jdk8u-dev/hotspot/src/share/vm/ci/ciConstant.hpp:29,
                 from /home/tester/jdk8u-dev/hotspot/src/share/vm/ci/ciArray.hpp:29,
                 from /home/tester/jdk8u-dev/hotspot/src/share/vm/precompiled/precompiled.hpp:33:
/home/tester/jdk8u-dev/hotspot/src/share/vm/ci/ciInstance.hpp: In function ‘static const TypeAryPtr* TypeAryPtr::make(TypePtr::PTR, const TypeAry*, ciKlass*, bool, int, int, const TypeOopPtr*, int)’:
/home/tester/jdk8u-dev/hotspot/src/share/vm/ci/ciInstance.hpp:68:12: note: in a call to non-static member function ‘ciKlass* ciInstance::java_lang_Class_klass()’
   68 |   ciKlass* java_lang_Class_klass();
      |            ^~~~~~~~~~~~~~~~~~~~~
```

**Comment by sxa** [comment]
> > > I have done some additional testing:

@zzambers Looks similar to what I'm seeing with GCC 11.2/aarch64, plus I still have the error from my [earlier comment](https://github.com/openjdk/jdk8u-dev/pull/357#issuecomment-2091050348) which seems to be in your log too.
Ref: https://github.com/adoptium/temurin-build/issues/3758#issuecomment-2293651827 which links to the full log in our jenkins instance.

**Comment by zzambers** [comment]
@gnu-andrew Actually I have found that that there are actually backports to fix `type.cpp` (they also include tests):
https://github.com/openjdk/jdk8u-dev/pull/536
and this one (see discussion in higher PR):
https://github.com/openjdk/jdk8u-dev/pull/552
(so compiler warnings seem to symptom of real issue)

Issue, I referenced earlier actually seems like duplicate:
https://github.com/openjdk/jdk8u-dev/pull/479
(I commented there)

**Pull request closed without merging**